pkg/vfs/download_store.go

```diff
@@ -0,0 +1,144 @@
+package vfs
+
+import (
+	"encoding/hex"
+	"sync"
+	"time"
+
+	"github.com/cozy/cozy-stack/pkg/crypto"
+)
+
+// A DownloadStore is essentially an object to store Archives & Files by keys
+type DownloadStore interface {
+	AddFile(f string) (string, error)
+	AddArchive(a *Archive) (string, error)
+	GetFile(k string) (string, error)
+	GetArchive(k string) (*Archive, error)
+}
+
+type fileRef struct {
+	Path      string
+	ExpiresAt time.Time
+}
+
+// downloadStoreTTL is the time an Archive stay alive
+const downloadStoreTTL = 1 * time.Hour
+const downloadStoreCleanInterval = 1 * time.Hour
+
+var storeStoreMutex = &sync.Mutex{}
+var storeStore map[string]*memStore
+
+func init() {
+	go cleanDownloadStoreInterval()
+}
+
+func cleanDownloadStoreInterval() {
+	for range time.Tick(downloadStoreCleanInterval) {
+		cleanDownloadStore()
+	}
+}
+
+func cleanStore(s *memStore) {
+	now := time.Now()
+	for k, f := range s.Files {
+		if now.After(f.ExpiresAt) {
+			delete(s.Files, k)
+		}
+	}
+	for k, a := range s.Archives {
+		if now.After(a.ExpiresAt) {
+			delete(s.Archives, k)
+		}
+	}
+}
+
+func cleanDownloadStore() {
+	storeStoreMutex.Lock()
+	defer storeStoreMutex.Unlock()
+	for i, s := range storeStore {
+		cleanStore(s)
+		if len(s.Files) == 0 && len(s.Archives) == 0 {
+			delete(storeStore, i)
+		}
+	}
+}
+
+// GetStore returns the DownloadStore for the given Instance
+func GetStore(domain string) DownloadStore {
+	storeStoreMutex.Lock()
+	defer storeStoreMutex.Unlock()
+	if storeStore == nil {
+		storeStore = make(map[string]*memStore)
+	}
+	store, exists := storeStore[domain]
+	if !exists {
+		store = &memStore{
+			Mutex:    &sync.Mutex{},
+			Archives: make(map[string]*Archive),
+			Files:    make(map[string]*fileRef),
+		}
+		storeStore[domain] = store
+	}
+	return store
+}
+
+type memStore struct {
+	Mutex    *sync.Mutex
```

Comment:
More infos in [this post](https://dave.cheney.net/2013/01/19/what-is-the-zero-value-and-why-is-it-useful).